### PR TITLE
Added compute shaders to the new pipelined renderer.

### PR DESCRIPTION
**Pull request by StarArawn**
The goal of this PR is not to add a high level API for GPU compute to bevy, rather to add a corner stone for compute, which will allow different people to think of high level API's that can be used to expose compute functionality to bevy users. This adds and exposes pipelines and a compute pass and all of the required stuff from wgpu to make "compute" functional from a very low level.

Note: This adds zero easy of use compute functionality such as a "compute node".

A list of major changes:

- `PipelineDescriptor` > `RenderPipelineDescriptor`
- `RenderContext`
  - `begin_pass` > `begin_render_pass`
  - new `begin_compute_pass`
  - new `create_compute_pipeline`
- new `ComputePass`
- new `ComputePipelineDescriptor`

### CODE REVIEW

File: crates/bevy_app/src/app.rs [hunk 1]
```diff
@@ -568,7 +568,11 @@ impl App {
         self
     }
 
-    pub fn add_sub_app(&mut self, app: App, f: impl Fn(&mut World, &mut App) + 'static) -> &mut Self {
+    pub fn add_sub_app(
```
Comment (StarArawn) [inline crates/bevy_app/src/app.rs:571]: `cargo fmt` changes to some files..

File: pipelined/bevy_wgpu2/src/render_resource_context.rs [hunk 2]
```diff
@@ -4,8 +4,8 @@ use crate::{
 };
 use bevy_render2::{
     pipeline::{
-        BindGroupDescriptor, BindGroupDescriptorId, BindingShaderStage, PipelineDescriptor,
-        PipelineId,
+        BindGroupDescriptor, BindGroupDescriptorId, BindingShaderStage, ComputePipelineDescriptor,
+        PipelineDescriptor, PipelineId,
```
Comment (StarArawn) [inline pipelined/bevy_wgpu2/src/render_resource_context.rs]: We might want to rename `PipelineDescriptor` to `RenderPipelineDescriptor`?

File: pipelined/bevy_wgpu2/src/compute_pass.rs [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+    pub wgpu_resources: WgpuResourceRefs<'a>,
+    pub pipeline_descriptor: Option<&'a ComputePipelineDescriptor>,
+}
+
+impl<'a> ComputePass for WgpuComputePass<'a> {
+    fn get_render_context(&self) -> &dyn RenderContext {
+        self.render_context
+    }
+
+    fn set_bind_group(
```
Comment (StarArawn) [inline pipelined/bevy_wgpu2/src/compute_pass.rs:23]: Would be nice if we could share this with render pass somehow.. perhaps pull it out into its own function?

File: crates/bevy_app/src/app.rs (hunk 1 above)
Comment (cart) [reply crates/bevy_app/src/app.rs:571]: proper formatting in a prototype? get that out of here :smile:

File: pipelined/bevy_wgpu2/src/render_resource_context.rs (hunk 2 above)
Comment (cart) [reply pipelined/bevy_wgpu2/src/render_resource_context.rs]: Yeah. That is both more descriptive and in-line with wgpu (which might ultimately replace this api).

File: pipelined/bevy_wgpu2/src/compute_pass.rs (hunk 3 above)
Comment (cart) [reply pipelined/bevy_wgpu2/src/compute_pass.rs:23]: yeah im down for that!

File: pipelined/bevy_wgpu2/src/render_resource_context.rs (hunk 2 above)
Comment (StarArawn) [reply pipelined/bevy_wgpu2/src/render_resource_context.rs]: Done!

File: pipelined/bevy_wgpu2/src/compute_pass.rs (hunk 3 above)
Comment (StarArawn) [reply pipelined/bevy_wgpu2/src/compute_pass.rs:23]: It got a little weird with a `Pass` enum, but I think it works fairly well?

Comment (cart) [reply pipelined/bevy_wgpu2/src/compute_pass.rs:23]: I'm guessing that gets optimized out, but it still feels a little weird to me. I actually think id prefer code duplication over that :)

Comment (StarArawn) [reply pipelined/bevy_wgpu2/src/compute_pass.rs:23]: I reversed the change.

**Comment by cart** [comment]
Oh no CI broke because formatting is broken. Now we need to run `cargo fmt`, push the changes, and wait for CI to finish!

**Pull request merged**